File: tools/webhook-verify/php/test_verify.php
<?php
/**
 * Smoke tests for the PHP webhook verify helper.
 *
 * Run as:  php test_verify.php
 * Exits 0 on success, 1 on any failure.
 */

require __DIR__ . '/verify.php';

$SECRET = 'test-secret-do-not-use';
$BODY   = '{"event":"webhook.test","occurredAt":"2026-05-01T00:00:00.000Z","payload":{"hello":"world"}}';

function good_sig($secret, $body) {
    return 'sha256=' . hash_hmac('sha256', $body, $secret);
}

$cases = [
    // [label, body, sig, secret, expected]
    ['valid sig',         $BODY,         good_sig($SECRET, $BODY), $SECRET, true],
    ['tampered body',     $BODY . '!',   good_sig($SECRET, $BODY), $SECRET, false],
    ['wrong secret',      $BODY,         good_sig($SECRET, $BODY), 'different-secret', false],
    ['missing prefix',    $BODY,         substr(good_sig($SECRET, $BODY), 7), $SECRET, false],
    ['empty sig header',  $BODY,         '', $SECRET, false],
    ['empty body, valid', '',            good_sig($SECRET, ''), $SECRET, true],
];

$failures = [];
foreach ($cases as $c) {
    [$label, $body, $sig, $secret, $expected] = $c;
    $actual = mc_verify_webhook($body, $sig, $secret);
    if ($actual !== $expected) {
        $failures[] = sprintf('%s: expected %s, got %s', $label, var_export($expected, true), var_export($actual, true));
    }
}

if (count($failures) > 0) {
    foreach ($failures as $f) {
        fwrite(STDERR, "FAIL: $f\n");
    }
    exit(1);
}

echo 'ok (' . count($cases) . " cases)\n";
exit(0);

File: tools/webhook-verify/python/test_verify.py
"""Smoke tests for the Python webhook verify helper.

Run as:  python test_verify.py
Exits 0 on success, 1 on any failure.
"""

import hmac
import hashlib
import sys

from verify import verify


SECRET = "test-secret-do-not-use"
BODY = (
    b'{"event":"webhook.test","occurredAt":"2026-05-01T00:00:00.000Z",'
    b'"payload":{"hello":"world"}}'
)


def good_sig(secret=SECRET, body=BODY):
    return "sha256=" + hmac.new(secret.encode(), body, hashlib.sha256).hexdigest()


def main():
    failures = []

    cases = [
        # (label, raw_body, sig_header, secret, expected_result)
        ("valid sig", BODY, good_sig(), SECRET, True),
        ("tampered body", BODY + b"!", good_sig(), SECRET, False),
        ("wrong secret", BODY, good_sig(), "different-secret", False),
        ("missing prefix", BODY, good_sig().replace("sha256=", ""), SECRET, False),
        ("empty sig header", BODY, "", SECRET, False),
        ("bytes secret", BODY, good_sig(), SECRET.encode(), True),
        ("str body", BODY.decode(), good_sig(), SECRET, True),
    ]

    for label, body, sig, secret, expected in cases:
        actual = verify(body, sig, secret)
        if actual != expected:
            failures.append(f"{label}: expected {expected}, got {actual}")

    if failures:
        for f in failures:
            print("FAIL:", f)
        sys.exit(1)
    print(f"ok ({len(cases)} cases)")


if __name__ == "__main__":
    main()
